tests/upgrade_test.go

```diff
@@ -0,0 +1,122 @@
+}
+
+// test upgrade single server rocksdb 3.3 -> 3.4
+func TestUpgradeSingleRocksDB33to34(t *testing.T) {
+	upgradeSubTest(t, api.DeploymentModeSingle, api.StorageEngineRocksDB, "3.3.0", "3.4")
+}
+
+// test upgrade resilient single server rocksdb 3.2 -> 3.3
+func TestUpgradeResilientSingleRocksDB32to33(t *testing.T) {
+	upgradeSubTest(t, api.DeploymentModeResilientSingle, api.StorageEngineRocksDB, "3.2.0", "3.3.3")
+}
+
+// test upgrade resilient single server mmfiles 3.3 -> 3.4
+func TestUpgradeResilientSingleMMFiles33to34(t *testing.T) {
+	upgradeSubTest(t, api.DeploymentModeResilientSingle, api.StorageEngineMMFiles, "3.3.0", "3.4")
+}
+
+// test upgrade cluster rocksdb 3.2 -> 3.3
+func TestUpgradeClusterRocksDB32to33(t *testing.T) {
+	upgradeSubTest(t, api.DeploymentModeCluster, api.StorageEngineRocksDB, "3.2.0", "3.3.3")
+}
+
+// test upgrade cluster mmfiles 3.3 -> 3.4
+func TestUpgradeClusterMMFiles33to34(t *testing.T) {
+	upgradeSubTest(t, api.DeploymentModeCluster, api.StorageEngineRocksDB, "3.3.0", "3.4")
+}
+
+func upgradeSubTest(t *testing.T, mode api.DeploymentMode, engine api.StorageEngine, fromVersion, toVersion string) error {
+	// check environment
+	longOrSkip(t)
+
+	k8sNameSpace := getNamespace(t)
+	k8sClient := mustNewKubeClient(t)
+	deploymentClient := kubeArangoClient.MustNewInCluster()
+
+	// Prepare deployment config
+	deploymentTemplate := newDeployment("test-upgrade-" + string(mode) + "-" + string(engine) + "-" + fromVersion + "to" + toVersion + "-" + uniuri.NewLen(4))
+	deploymentTemplate.Spec.Mode = api.NewMode(mode)
+	deploymentTemplate.Spec.StorageEngine = api.NewStorageEngine(engine)
+	deploymentTemplate.Spec.TLS = api.TLSSpec{} // should auto-generate cert
+	deploymentTemplate.Spec.Image = util.NewString("arangodb/arangodb:" + fromVersion)
+	deploymentTemplate.Spec.SetDefaults(deploymentTemplate.GetName()) // this must be last
+
+	// Create deployment
+	deployment, err := deploymentClient.DatabaseV1alpha().ArangoDeployments(k8sNameSpace).Create(deploymentTemplate)
+	if err != nil {
+		t.Fatalf("Create deployment failed: %v", err)
+	}
+
+	// Wait for deployment to be ready
+	deployment, err = waitUntilDeployment(deploymentClient, deploymentTemplate.GetName(), k8sNameSpace, deploymentHasState(api.DeploymentStateRunning))
+	if err != nil {
+		t.Fatalf("Deployment not running in time: %v", err)
+	}
+
+	arangoDeploymentHealthy(t, deployment, k8sClient)
+
+	// Try to change image version
+	deployment, err = updateDeployment(deploymentClient, deploymentTemplate.GetName(), k8sNameSpace,
+		func(spec *api.DeploymentSpec) {
+			spec.Image = util.NewString("arangodb/arangodb:" + toVersion)
+		})
+	if err != nil {
+		t.Fatalf("Failed to upgrade the Image from version : " + fromVersion + " to version: " + toVersion)
+	}
+
+	arangoDeploymentHealthy(t, deployment, k8sClient)
+
+	ctx := context.Background()
+	DBClient := mustNewArangodDatabaseClient(ctx, k8sClient, deployment, t)
```

Comment:
I think this test must be done on all servers.
The current implementation of the test can succeed even f not all servers are actually upgraded.

Reply:
Also, you put this test in an `op := func() error {... ` with a `Retry`: "Wait until all versions are correct"
Reason: It could be that the change of image is saved, the cluster is still healthy and the version check test
starts before any upgrading has actually taken place.